### PR TITLE
backup: support kms decryption secret

### PR DESCRIPTION
**Pull request by shuijing198799**
### What problem does this PR solve? 
support kms decryption secret in tidb-operator
### What is changed and how does it work?
add KMS field in backup CRD, it should set to be true if the secret has been encrypted by KMS
### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)

Code changes

 - Has Go code change

### Does this PR introduce a user-facing change?:

```release-note
Add `spec.useKMS` field in `Backup`, `Restore` and `BackupSchedule` custom resources. Set it to `true` if you want to decrypt the secrets encrypted with KMS.
```

### CODE REVIEW

File: pkg/backup/util/util.go [hunk 1]
```diff
@@ -226,9 +239,16 @@ func GenerateTidbPasswordEnv(ns, name, tidbSecretName string, secretLister corel
...
+	if useKMS {
+		passwordKey = fmt.Sprintf("%s_%s_%s", constants.KMSSecretPrefix, constants.BackupManagerEnvVarPrefix, strings.ToUpper(constants.TidbPasswordKey))
+	} else {
+		passwordKey = fmt.Sprintf("%s_%s", constants.BackupManagerEnvVarPrefix, strings.ToUpper(constants.TidbPasswordKey))
+	}
+
 	certEnv = []corev1.EnvVar{
 		{
-			Name: fmt.Sprintf("%s_%s", constants.BackupManagerEnvVarPrefix, strings.ToUpper(constants.TidbPasswordKey)),
+			Name: passwordKey,
```
Comment (DanielZhangQD) [inline pkg/backup/util/util.go]: backup-manager will get the password from the env, the code there may need to update, for example, https://github.com/pingcap/tidb-operator/blob/master/cmd/backup-manager/app/backup/manager.go#L67

**Comment by gregwebs** [comment]
I think this should be switched to sops (it does have exec-env) so that we can support multiple clouds.

**Comment by shuijing198799** [comment]
> I think this should be switched to sops (it does have exec-env) so that we can support multiple clouds.

It should support by DBaaS first, and then I'll modify the operator code

**Comment by shuijing198799** [comment]
> I think this should be switched to sops (it does have exec-env) so that we can support multiple clouds.

Maybe it will change to sop after support GCP ?

File: pkg/backup/util/util.go [hunk 2]
```diff
@@ -87,6 +87,8 @@ func GenerateS3CertEnvVar(s3 *v1alpha1.S3StorageProvider) ([]corev1.EnvVar, stri
 			Value: s3.StorageClass,
 		},
 	}
+	// KMS has to use with IAM for get credentail to decryption the code
```
Comment (DanielZhangQD) [inline pkg/backup/util/util.go]: Please update the description here.

File: pkg/backup/util/util.go [hunk 3]
```diff
@@ -39,7 +39,7 @@ func CheckAllKeysExistInSecret(secret *corev1.Secret, keys ...string) (string, b
 }
 
 // GenerateS3CertEnvVar generate the env info in order to access S3 compliant storage
-func GenerateS3CertEnvVar(s3 *v1alpha1.S3StorageProvider) ([]corev1.EnvVar, string, error) {
+func GenerateS3CertEnvVar(useKMS bool, s3 *v1alpha1.S3StorageProvider) ([]corev1.EnvVar, string, error) {
```
Comment (DanielZhangQD) [inline pkg/backup/util/util.go]: useKMS is not used in this function.

File: pkg/backup/util/util.go [hunk 4]
```diff
@@ -87,6 +87,8 @@ func GenerateS3CertEnvVar(s3 *v1alpha1.S3StorageProvider) ([]corev1.EnvVar, stri
 			Value: s3.StorageClass,
 		},
 	}
+	// KMS has to use with IAM for get credentail to decryption the code
+	// so if use KMS, the secretKey and accesssKey can be ignore
 	if s3.SecretName != "" {
 		envVars = append(envVars, []corev1.EnvVar{
 			{
```
Comment (DanielZhangQD) [inline pkg/backup/util/util.go]: No need to consider the case where accessKeys are encrypted by KMS?

File: pkg/backup/util/util.go (hunk 3 above)
Comment (shuijing198799) [reply pkg/backup/util/util.go]: done

File: pkg/backup/util/util.go (hunk 2 above)
Comment (shuijing198799) [reply pkg/backup/util/util.go]: done

File: pkg/backup/util/util.go (hunk 4 above)
Comment (shuijing198799) [reply pkg/backup/util/util.go]: though use rarely use like this, we can't don't allow them do like this

File: pkg/backup/util/util.go [hunk 5]
```diff
@@ -152,7 +152,7 @@ func GenerateGcsCertEnvVar(gcs *v1alpha1.GcsStorageProvider) ([]corev1.EnvVar, s
 }
 
 // GenerateStorageCertEnv generate the env info in order to access backend backup storage
-func GenerateStorageCertEnv(ns string, provider v1alpha1.StorageProvider, secretLister corelisters.SecretLister) ([]corev1.EnvVar, string, error) {
+func GenerateStorageCertEnv(ns string, useKMS bool, provider v1alpha1.StorageProvider, secretLister corelisters.SecretLister) ([]corev1.EnvVar, string, error) {
```
Comment (DanielZhangQD) [inline pkg/backup/util/util.go]: useKMS is not used.

Comment (shuijing198799) [reply pkg/backup/util/util.go]: done

File: pkg/backup/constants/constants.go [hunk 6]
```diff
@@ -55,4 +55,7 @@ const (
 
 	// ServiceAccountCAPath is where is CABundle of serviceaccount locates
 	ServiceAccountCAPath = "/var/run/secrets/kubernetes.io/serviceaccount/ca.crt"
+
+	// KMS secret env prefix
+	KMSSecretPrefix = "KMS_ENCRYPTED"
```
Comment (cofyc) [inline pkg/backup/constants/constants.go]: how this is being used? any docs or links about this env?

File: images/tidb-backup-manager/entrypoint.sh [hunk 7]
```diff
@@ -51,33 +51,40 @@ else
 fi
 
 BACKUP_BIN=/tidb-backup-manager
+if [[ -n "${AWS_DEFAULT_REGION}"]]; then
```
Comment (cofyc) [inline images/tidb-backup-manager/entrypoint.sh]: this env `AWS_DEFAULT_REGION` is always injected into pods on eks?

File: pkg/backup/constants/constants.go (hunk 6 above)
Comment (shuijing198799) [reply pkg/backup/constants/constants.go]: the dbaas encrpytion secret with KMS so the  secret mount to pod should decryption with [shush](https://github.com/realestate-com-au/shush) this prefix means this env should decryption

File: images/tidb-backup-manager/entrypoint.sh (hunk 7 above)
Comment (shuijing198799) [reply images/tidb-backup-manager/entrypoint.sh]: this AWS_DEFAULT_REGION only use in useKMS is on

**Pull request merged**